### ISSUE TITLE
Does SetMagickResourceLimit(ThreadResource, limit_) impose restrictions on functions accelerated by OpenCL?

### ISSUE THREAD

**Issue by eakey** [ImageMagick/ImageMagick#7342]
### ImageMagick version

7.1.1-33

### Operating system

Windows

### Operating system, version and so on

vs2017 v141 windows sdk 8.1

### Description

I examined the ModulateImage function in the enhance.c file, which has an OpenCL-accelerated version named AccelerateModulateImage. I found that the SetMagickResourceLimit(ThreadResource, limit_) setting does not appear to be effective for functions that support OpenCL acceleration. If OpenCL runs on the CPU, would it potentially utilize all CPU cores to their full capacity?

### Steps to Reproduce

none

**Comment by dlemstra** [comment]
That is correct and a limitation of the OpenCL library. It will always use all cores when running on CPU. We could add an extra check for this that when an CPU OpenCL device is used and the thread limit is set to 1 we will not allow using this device.

**Comment by eakey** [comment]
> That is correct and a limitation of the OpenCL library. It will always use all cores when running on CPU. We could add an extra check for this that when an CPU OpenCL device is used and the thread limit is set to 1 we will not allow using this device.

Is it the case that when OpenCL is running on the CPU, and SetMagickResourceLimit(ThreadResource, 2) is in effect, with the invoked function having an OpenCL implementation, the CPU usage will rise to 100%?

**Comment by dlemstra** [comment]
I don't know if your CPU usage will reach to 100% but all cores will be used. And I just realized you can already prevent the CPU being used for OpenCL operation by settings `MAGICK_OCL_DEVICE` to `GPU` so I don't think we will change our code to look at the `ThreadResource` limitation.

**Comment by eakey** [comment]
> SetMagickResourceLimit(ThreadResource, 2)

My application environment is restricted to using CPU-based OPENCL exclusively. Although an interface like SetMagickResourceLimit(ThreadResource, 2) is supposed to limit the number of CPU cores used, it fails to meet expectations in certain scenarios and may lead to issues, such as CPU resource depletion, under specific circumstances. Being unfamiliar with OPENCL and having recently started researching, I've found that there should be methods to control the number of CPU cores utilized within the OPENCL framework.

**Comment by dlemstra** [comment]
As I said before the api of the OpenCL library that is being used does not allow limiting the number of cores that are being used by an OpenCL CPU operation. This is out of our control so we cannot do much about this. OpenCL is opt-in only so if this is really an issue for you I would advise you to disable OpenCL operations.

**Comment by eakey** [comment]
> As I said before the api of the OpenCL library that is being used does not allow limiting the number of cores that are being used by an OpenCL CPU operation. This is out of our control so we cannot do much about this. OpenCL is opt-in only so if this is really an issue for you I would advise you to disable OpenCL operations.

I've observed that these functions which support OPENCL acceleration do not utilize SIMD (Single Instruction, Multiple Data) techniques when OPENCL is disabled, resulting in a significant performance degradation when OPENCL is turned off.

**Comment by dlemstra** [comment]
Don't forget that you are also getting a performance improvement because you are using all cores of the CPU.

**Comment by eakey** [comment]
> Don't forget that you are also getting a performance improvement because you are using all cores of the CPU.

No, I am limited to using only 2 cores, and without SIMD technology, the performance would be worse than that of a single core utilizing SIMD. Furthermore, if I enable OPENCL, even if I restrict the usage to 2 cores, ImageMagick will still utilize all available cores.

**Comment by eakey** [comment]
That's why I was considering if setting SetMagickResourceLimit(ThreadResource, 2) could also apply to OPENCL. I'm currently looking into how to limit CPU core usage within OPENCL. When OPENCL runs on the CPU, it essentially leverages the CPU's SIMD capabilities for acceleration.

**Comment by dlemstra** [comment]
Closing this because we don't want to add this extra functionality.
